NetSim: More reliable address assignment

islemaster opened the pull request:
We were waiting for way too many server requests when connecting to a router.  In recent testing, we found out that this is a particularly easy process to interrupt with a shard reset (see [this Honeybadger incident](https://app.honeybadger.io/projects/3240/faults/14988376)).

There's a lot of clean-up here, but this is the end result (each step requires an AJAX request):

**Router connection process before**
1. Insert a wire row (only nodeIDs filled in).
2. Read wire table to make sure the router is not over capacity.  If it is, stop and disconnect.
3. Update the router row (used to be necessary to update router status).
4. Read the wire table again to find an unused address (No protection against address collisions).
5. Update the wire row to add addresses and hostnames.  If the update fails, stop and disconnect.

**Router connection process now**
1. Insert a wire row (nodeIDs, addresses and hostnames filled in, using cached wire data).
2. Read the wire table to make sure the router is not over capacity and check for address collisions.  If either occurs, stop and disconnect.

**Other improvements along the way**
- `NetSimLocalClientNode` does not cache `myWire` since it can be calculated from the cached state.  It's a little less efficient, but also less stateful and error-prone.
- `NetSimRouterNode`'s `getConnections` and `countConnections` functions used to trigger server requests, but are now synchronous and use the cache.  This shouldn't be a problem for router operations, and should help a lot since relatively simple things like getting router status to display in the lobby were using these functions.
- Added tests to ensure the API's behavior when a row is not found on update and delete operations.

Hoping this will finish [Pivotal #101529076](https://www.pivotaltracker.com/story/show/101529076).

Hamms commented on apps/src/netsim/NetSimRouterNode.js:
```diff
@@ -925,85 +917,76 @@ var contains = function (haystack, needle) {
-    }).map(function (wire) {
-      return wire.localAddress;
-    });
+NetSimRouterNode.prototype.getRandomAvailableClientAddress = function () {
```
Why random and not next? Seems like we're doing a lot more work for that random functionality; is it really that desirable?

islemaster replied:
Explicit feature request from the ed team - in lessons designed to motivate DNS it is too easy to guess hidden user addresses when they are sequential, especially in small groups.

Initially implemented in PR #2540

Hamms replied:
:+1:

Hamms commented on apps/test/netsim/NetSimLocalClientNode.js:
```diff
@@ -284,6 +299,66 @@ describe("NetSimLocalClientNode", function () {
+        wireRow = testLocalNode.makeWireRowForConnectingTo(routerNode);
+      });
+
+      it ("Sets own entity ID to localNodeID", function () {
```
is this description backwards?

islemaster replied:
Yup.

Hamms commented on apps/test/netsim/NetSimRouterNode.js:
```diff
@@ -373,45 +373,39 @@ describe("NetSimRouterNode", function () {
+      NetSimWire.create(testShard, {
+            localNodeID: 0,
+            remoteNodeID: routerA.entityID
+          } + 1, function () {});
```
I'm assuming this should be
```
  remoteNodeID: routerA.entityID + 1
}, function () {});
```

If this test was passing without errors, I'm a little worried; should we be doing some more verification somewhere of valid data?

islemaster replied:
Great catch!

I updated `NetSimWire.create()` with argument validation so that this test (with typo intact) fails, then fixed the typo.

Hamms commented on apps/test/netsim/NetSimRouterNode.js:
```diff
@@ -373,45 +373,39 @@ describe("NetSimRouterNode", function () {
-      });
       // Only get the one wire back.
-      assertEqual(wires.length, 1);
+      assertEqual(1, routerA.getConnections().length);
```
I'd also like to verify that the one wire we got back is the expected wire

Hamms commented on apps/test/netsim/NetSimRouterNode.js:
```diff
@@ -427,7 +421,11 @@ describe("NetSimRouterNode", function () {
+        NetSimWire.create(testShard, {
+          localNodeID: wireID,
+          remoteNodeID: routerA.entityID,
+          localAddress: wireID.toString(10)
```
What's the purpose of this?

islemaster replied:
As of this change, routers reject connections when they detect an address collision.  Here we are making sure each wire has a different address so no collisions occur.

Hamms commented on apps/src/netsim/NetSimRouterNode.js:
```diff
@@ -689,16 +681,34 @@ NetSimRouterNode.prototype.getStatus = function () {
+NetSimRouterNode.prototype.getConnectedNodeNames_ = function () {
+  var cachedNodeRows = this.shard_.nodeTable.readAll();
+  return this.getConnections()
+      .map(function (wire) {
```
I'd much prefer to see `return this.getConnections().map(function (wire) {` all on one line here; makes it much harder to accidentally see `return this.getConnections()`  and assume that's all that's being returned.

Hamms commented on apps/src/netsim/NetSimRouterNode.js:
```diff
@@ -849,43 +859,25 @@ NetSimRouterNode.prototype.setMemory = function (newMemory) {
-    .fail(function (err) {
-        onComplete(err, []);
+  return shard.wireTable.readAll()
+      .filter(function (wireRow) {
```
likewise, `return shard.wireTable.readAll().filter(function (wireRow) {` all on one line here is much harder to misread

islemaster commented:
Great feedback!  I think I've addressed all your comments so far.  Please scan the changes and approve/deny when you can; I'm headed out for the weekend though, so no rush.

Hamms commented:
LGTM!
